File: LEETCODE_diameter_of_binary_tree.rb
# Time complexity O(n)
# Space complexity O(height of btree)
def diameter_of_binary_tree(root)
    @res = 0
    dfs(root)
    return @res
end

def dfs(node)
    return 0 unless node

    left = dfs(node.left)
    right = dfs(node.right)
    @res = [@res, left + right].max

    return 1 + [left, right].max
end

File: LEETCODE_diameter_of_binary_tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int res = 0;

    public int DiameterOfBinaryTree(TreeNode root) {
        Helper(root);
        return res;
    }

    private int Helper(TreeNode node)
    {
        if (node == null) return 0;

        int leftCount = Helper(node.left);
        int rightCount = Helper(node.right);
        int height = 1 + Math.Max(leftCount, rightCount);
        this.res = Math.Max(res, leftCount + rightCount);
        return height;
    }
}
